api/src/context.rs

```diff
@@ -136,6 +130,109 @@ impl Context {
+                return Ok(None);
+            }
+        };
+
+        let txn_with_proof = self
+            .db
+            .get_transaction_by_version(start, ledger_version, false)?;
+
+        // Retrieve block timestamp and hash
+        let timestamp;
+        let block_hash;
+        use aptos_types::transaction::Transaction::*;
+        match &txn_with_proof.transaction {
+            GenesisTransaction(_) => {
+                timestamp = 0;
+                block_hash = txn_with_proof.proof.transaction_info.transaction_hash();
+            }
+            BlockMetadata(inner) => {
+                timestamp = inner.timestamp_usecs();
+                block_hash = inner.id();
+            }
+            _ => {
+                return Err(anyhow!(
+                    "Failed to retrieve BlockMetadata or Genesis transaction"
+                ));
+            }
+        }
+
+        // If timestamp is 0, it's the genesis transaction, and we can stop now
+        if timestamp == 0 {
+            return Ok(Some(BlockInfo {
+                block_height: 0,
+                start_version: start,
+                end_version: end,
+                block_hash: block_hash.into(),
+                block_timestamp: timestamp,
+                num_transactions: end.saturating_sub(start).saturating_add(1) as u16,
+            }));
+        }
+
+        // Retrieve block height from the transaction outputs
+        let height_id = ident_str!("height");
+        let block_metadata_type = move_deps::move_core_types::language_storage::StructTag {
+            address: AccountAddress::ONE,
+            module: ident_str!("Block").into(),
+            name: ident_str!("BlockMetadata").into(),
+            type_params: vec![],
+        };
+
+        let resolver = self.move_resolver()?;
+        let converter = resolver.as_converter();
+        let txn = self.get_transaction_by_version(start, ledger_version)?;
+
+        // Parse the resources and find the block metadata resource update
+        let maybe_block_height = txn.changes.iter().find_map(|(key, op)| {
+            if let StateKey::AccessPath(path) = key {
+                if let Path::Resource(typ) = path.get_path() {
+                    // If it's block metadata, we can convert it to get the block height
+                    // And it must be the root address
+                    if path.address == aptos_root_address() && typ == block_metadata_type {
+                        if let WriteOp::Value(value) = op {
+                            if let Ok(mut resource) = converter.try_into_resource(&typ, value) {
+                                if let Some(value) = resource.data.0.remove(height_id) {
+                                    if let Ok(height) = serde_json::from_value::<U64>(value) {
+                                        return Some(height.0);
+                                    }
+                                }
+                            }
```

Comment:
Shall we define a Rust struct corresponding to the on chain struct and bsc deserialize the bytes into it? Using the converter will be slow.

Also @zekun000 , we can't make the height into the txn itself or the event?

Reply:
we can add height to the event